modules/core/src/main/java/org/apache/ignite/internal/managers/encryption/GridEncryptionManager.java

```diff
@@ -734,18 +884,309 @@ private void writeToMetaStore(int grpId, byte[] encGrpKey) {
+                return new GridFinishedFuture<>(new IgniteException("Master key change was rejected. Master " +
+                    "key digest consistency check failed. Make sure that a new master key is the same at " +
+                    "all server nodes [nodeId=" + ctx.localNodeId() + ']'));
+            }
+        }
+        catch (Exception e) {
+            return new GridFinishedFuture<>(new IgniteException("Master key change was rejected [nodeId=" +
+                ctx.localNodeId() + ']', e));
+        }
+
+        return new GridFinishedFuture<>(new MasterKeyChangeResult());
+    }
+
+    /**
+     * Starts master key change process if there are no errors.
+     *
+     * @param id Request id.
+     * @param res Results.
+     * @param err Errors.
+     */
+    private void finishPrepareMasterKeyChange(UUID id, Map<UUID, MasterKeyChangeResult> res, Map<UUID, Exception> err) {
+        if (!err.isEmpty()) {
+            if (pendingMasterKey != null && pendingMasterKey.requestId().equals(id))
+                pendingMasterKey = null;
+
+            doneMasterKeyChangeFuture(id, err);
+        }
+        else if (isCoordinator())
+            finishProc.start(id, pendingMasterKey);
+    }
+
+    /**
+     * Changes master key.
+     *
+     * @param req Request.
+     * @return Result future.
      */
-    private Collection<byte[]> createKeys(int keyCnt) {
-        if (keyCnt == 0)
-            return Collections.emptyList();
+    private IgniteInternalFuture<MasterKeyChangeResult> masterKeyChange(PendingMasterKey req) {
+        if (pendingMasterKey == null || !pendingMasterKey.equals(req))
+            return new GridFinishedFuture<>(new IgniteException("Unknown master key change was rejected."));
+
+        if (!ctx.state().clusterState().active()) {
+            pendingMasterKey = null;
+
+            return new GridFinishedFuture<>(new IgniteException("Master key change was rejected. " +
+                "The cluster is inactive."));
+        }
+
+        if (!ctx.clientNode())
+            changeMasterKeyAndReencryptGroupKeys(decryptKeyName(req.encKeyName()));
 
-        List<byte[]> encKeys = new ArrayList<>(keyCnt);
+        pendingMasterKey = null;
 
-        for(int i=0; i<keyCnt; i++)
-            encKeys.add(getSpi().encryptKey(getSpi().create()));
+        lastChangedMasterKeyDigest = req.digest();
 
-        return encKeys;
+        return new GridFinishedFuture<>(new MasterKeyChangeResult());
+    }
+
+    /**
+     * Finishes master key change.
+     *
+     * @param id Request id.
+     * @param res Results.
+     * @param err Errors.
+     */
+    private void finishMasterKeyChange(UUID id, Map<UUID, MasterKeyChangeResult> res, Map<UUID, Exception> err) {
+        doneMasterKeyChangeFuture(id, err);
+    }
+
+    /**
+     * @param reqId Request id.
+     * @param err Exception.
+     */
+    private void doneMasterKeyChangeFuture(UUID reqId, Map<UUID, Exception> err) {
```

Comment:
Let's rename this to the `completeMasterKeyChangeFuture`

Reply:
Done.